frontends/PyCDE/src/pycde/matrix.py

```diff
@@ -0,0 +1,264 @@
+  # (len(dims) == 0).
+  dims: list
+  dtype: type
+
+  @property
+  def num_dims(self):
+    return len(self.dims)
+
+  @property
+  def type(self):
+    if len(self.dims) != 0:
+      return dim(self.dtype, *self.dims)
+    else:
+      return self.dtype
+
+  def __str__(self):
+    return str(self.type)
+
+
+class Matrix(np.ndarray):
+  """
+  A PyCDE Matrix serves as a Numpy view of a multidimensional CIRCT array (ArrayType).
+  The Matrix ensures that all assignments to itself have been properly converted
+  to conform with insertion into the numpy array (circt_to_arr).
+  Once filled, a user can treat the Matrix as a numpy array.
+  The underlying CIRCT array is not materialized until to_circt is called.
+  """
+
+  def __array_finalize__(self, obj):
+    # Ensure Matrix-class attributes are propagated upon copying.
+    # see https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_finalize__
+    if obj is not None:
+      self.__dict__ = obj.__dict__
+
+  def __new__(cls,
+              shape: list = None,
+              dtype=None,
+              name: str = None,
+              from_value=None) -> None:
+    """Construct a matrix with the given shape and dtype.
+    This is a __new__ function since np.ndarray does not have an __init__ function.
+      Args:
+        shape: A tuple of integers representing the shape of the matrix.
+        dtype: the inner type of the matrix. This is a PyCDE type - the Numpy
+               matrix contains 'object'-typed values.
+      """
+
+    if bool(from_value) and bool(shape or dtype):
+      raise ValueError(
+          "Must specify either shape and dtype, or initialize from a value, but not both."
+      )
+
+    if from_value is not None:
+      if not isinstance(from_value, ListValue):
+        raise ValueError("from_value must be a ListValue")
+      shape = from_value.type.shape
+      dtype = from_value.type.inner_type
+      name = from_value.name
+
+    # Initialize the underlying np.ndarray
+    self = np.ndarray.__new__(cls, shape=shape, dtype=object)
+
+    if name is None:
+      name = "matrix"  # todo: require name?
+    self.name = name
+    self.pycde_dtype = dtype
+
+    # The SSA value of this matrix after it has been materialized.
+    # Once set, the matrix is immutable.
+    self.circt_output = None
+
+    if from_value is not None:
+      # PyCDE and numpy do not play nicely when doing np.arr(self.circt_to_arr(...))
+      # but individual assignments work.
+      target_shape = self.target_shape_for_idxs(0)
+      value_arr = self.circt_to_arr(from_value, target_shape)
+      for i, v in enumerate(value_arr):
+        super().__setitem__(self, i, v)
+
+    return self
+
+  @lru_cache
+  def get_constant(self, value: int, width: int = 32):
+    """ Get an IR constant backed by a constant cache."""
+    return hw.ConstantOp(ir.IntegerType.get_signless(width), value)
+
+  def circt_to_arr(self, value, target_shape):
```

Comment:
Should this be a static method? Also, it doesn't look like it's intended to be used externally, so it should have a '_' prefix.